aspnet/fundamentals/file-providers/sample/src/FileProviderSample/Controllers/HomeController.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.FileProviders;
+
+namespace FileProviderSample.Controllers
+{
+    public class HomeController : Controller
+    {
+        private readonly IFileProvider _fileProvider;
+        private static string _dateFileLastChanged = "Never";
+
+        public HomeController(IFileProvider fileProvider)
+        {
+            _fileProvider = fileProvider;
+        }
+
+        public IActionResult Index()
+        {
+            var contents = _fileProvider.GetDirectoryContents("");
+            return View(contents);
+        }
+
+        public IActionResult Watch()
```

Comment:
>  fires off twice on the first change

Hmm, I didn't see this problem on macOS. Might be an issue?

>  I don't like the recursive nature of what I have

Here's a more complicated, but also not-recursive version:
```c#
    public class Program
    {
        public static PhysicalFileProvider PhysicalFileProvider = new PhysicalFileProvider(Directory.GetCurrentDirectory());
        public static void Main(string[] args)
        {
            Console.WriteLine("Monitoring foo.txt for changes...");
            while (true)
            {
                MainAsync().GetAwaiter().GetResult();
            }
        }

        private static async Task MainAsync()
        {
            IChangeToken token = PhysicalFileProvider.Watch("foo.txt");
            var tcs = new TaskCompletionSource<object>();
            token.RegisterChangeCallback(state => ((TaskCompletionSource<object>)state).TrySetResult(null), tcs);
            await tcs.Task.ConfigureAwait(false);
            Console.WriteLine("foo.txt changed");
        }
    }
```